File: trunk/cing/python/cing/Scripts/iCingRobot.py
# Script for testing of FileUpload at the CGI server and the other commands at the main iCing server.
# Run: python -u $CINGROOT/python/cing/Scripts/iCingRobot.py
# There is a unit test at: cing.Scripts.test.testiCingRobot

from cing import cingDirTestsData
from cing.Libs.NTutils import * #@UnusedWildImport
import mimetools
import urllib2

FORM_ACCESS_KEY = "AccessKey"
FORM_USER_ID = "UserId"
FORM_UPLOAD_FILE_BASE = "UploadFile"
FORM_ACTION = "Action"
FORM_ACTION_RUN = "Run"
FORM_ACTION_SAVE = "Save"
FORM_ACTION_STATUS = "Status"
FORM_ACTION_PROJECT_NAME = "ProjectName"
FORM_ACTION_PURGE = "Purge"
FORM_ACTION_LOG = "Log"

RESPONSE_EXIT_CODE = 'ExitCode'
RESPONSE_SUCCESS = 'Success'
RESPONSE_RESULT = 'Result'
RESPONSE_DONE = 'done'

#DEFAULT_URL = 'http://nmr.cmbi.ru.nl' # production without https security turned on
DEFAULT_URL = 'https://nmr.cmbi.ru.nl' # production with https security turned on
#DEFAULT_URL = 'http://localhost' # local tomcat instance
#DEFAULT_URL = 'http://localhost' # local gwt embedded tomcat instance
DEFAULT_RPC_PORT = ''
#DEFAULT_RPC_PORT = ':8080'

DEFAULT_URL_PATH = 'icing' # ?? lower case?
#DEFAULT_URL_PATH = 'cing.iCing' # use for gwt embedded tomcat

rpcUrl=DEFAULT_URL+"/icing/serv/iCingServlet"
#    rpcUrl=DEFAULT_URL+DEFAULT_RPC_PORT+'/'+DEFAULT_URL_PATH+"/serv/iCingServlet" # TODO make this line working and not the below line!
#    rpcUrl=DEFAULT_URL+DEFAULT_RPC_PORT+'/'+DEFAULT_URL_PATH+"/iCingServlet"

nTmessage("rpcUrl: " + rpcUrl)

def getResultUrls(credentials, entryId, url=None):

    userId = credentials[0][1]
    accessKey = credentials[1][1]

    resultBaseUrl = os.path.join(url, 'tmp/cing', userId, accessKey)
    resultHtmlUrl = os.path.join(resultBaseUrl, entryId + ".cing", "index.html")
    resultLogUrl = os.path.join(resultBaseUrl, "cingRun.log")
    resultZipUrl = os.path.join(resultBaseUrl, entryId + "_CING_report.zip")

    return resultBaseUrl, resultHtmlUrl, resultLogUrl, resultZipUrl

def sendRequest(url, fields, files):
    """Function to send form fields and files to a given URL.
    """

    contentType, bodyData = encodeForm(fields, files)

    headerDict = {'User-Agent': 'anonymous',
                  'Content-Type': contentType,
                  'Content-Length': str(len(bodyData))
                  }

    request = urllib2.Request(url, bodyData, headerDict)
    response = urlOpen(request)

    if not response:
      return

    jsonTxt = response.read()

    result = _processResponse(jsonTxt)
    if result.get(RESPONSE_EXIT_CODE) != RESPONSE_SUCCESS:
        msg  = 'Request not successful. Action was: %s' % fields
        msg += ' Response was: %s' % jsonTxt
        nTwarning('Failure %s' % msg)
        return

    return result

def urlOpen(request):

    try:
        response = urllib2.urlopen(request)

    except urllib2.URLError, e:
        if hasattr(e, 'reason'):
            if isinstance(request, urllib2.Request):
              url = request.get_full_url()
            else:
              url = request

            msg = 'Connection to server URL %s failed with reason:\n%s' % (url, e.reason)

        elif hasattr(e, 'code'):
            msg =    'Server request failed and returned code:\n%s' % e.code

        else:
            msg = 'Server totally barfed with no reason or fail code'

        nTwarning('Failure %s'% msg)
        return

    return response

def _processResponse(text):
    """Convert the strings the iCingServer sends back into Python
       When Python 2.6 is the CCPn standards this should use the JSON
       libraries.
    """

    text = text[2:-2]

    dataDict = {}

    for pair in text.split('","'):
      data = pair.split('":"')

      if len(data) == 2:
        key , value = data
        dataDict[key] = value
      else:
        print "Trouble",  pair


    return dataDict

#########################################################################################
# Initial code from http://www.voidspace.org.uk/python/cgi.shtml#upload                                                #
#########################################################################################

BOUNDARY = mimetools.choose_boundary()

def encodeForm(fields, files=None, lineSep='\r\n',
               boundary='-----'+BOUNDARY+'-----'):
    """Function to encode form fields and files so that they can be sent to a URL"""

    if not files:
        files = []

    lines = []
    if isinstance(fields, dict):
        fields = fields.items()


    for (key, fileName, value) in files:
        #fileType = mimetypes.guess_type(fileName)[0] or 'application/octet-stream'
        fileType = 'application/octet-stream'

        lines.append('--' + boundary)
        lines.append('Content-Disposition: form-data; name="%s"; filename="%s"' % (key, fileName))
        if not fileType.startswith('text'):
            lines.append('Content-Transfer-Encoding: binary')
        lines.append('Content-Type: %s' % fileType)
        #lines.append('Content-Length: %s\r\n' % str(len(value)))
        lines.append('')
        lines.append(value)

    for (key, value) in fields:
        lines.append('--' + boundary)
        lines.append('Content-Disposition: form-data; name="%s"' % key)
        lines.append('')
        lines.append(value)


    lines.append('--' + boundary + '--')
    lines.append('')

    bodyData = lineSep.join(lines)
    contentType = 'multipart/form-data; boundary=%s' % boundary

    return contentType, bodyData


def iCingRobot():
#    nTwarning("Expect errors without a server up and running.")
    nTmessage("Firing up the iCing robot; aka example interface to CING")

    ## queries possible; do one at a time going down the list.
    ## After the run is started the status will let you know if the run is finished
    ## The log will show what the server is doing at any one time.
    doSave  = 1 # Upload to iCing and show derived urls
    doRun   = 0 # Start the run in Nijmegen
    doStatus= 0 # Find out if the run finished
    doLog   = 0 # Get the next piece of log file (may be empty)
    doPname = 0 # Get the project name back. This is the entryId below.
    doPurge = 0 # Remove data from server again.

    # User id should be a short id (<without any special chars.)
#    user_id = os.getenv("USER", "UnknownUser")
    user_id = "iCingRobot"
#    access_key = "123456"
    access_key = getRandomKey() # Use a different one in a production setup.

    entryId = '1brv' # 68K, smallest for quick testing.
#    entryId = 'gb1' # only included in xplor variant as single model.

    # Select one of the types by uncommenting it
    inputFileType = 'CCPN'
#    inputFileType = 'PDB'
#    inputFileType = 'XPLOR'

    ccpnFile = os.path.join(cingDirTestsData, "ccpn", entryId + ".tgz")
    pdbFile = os.path.join(cingDirTestsData, "pdb", entryId, 'pdb' + entryId + ".ent")
    xplorFile = os.path.join(cingDirTestsData, "xplor", entryId, entryId + ".pdb")

    inputFile = ccpnFile
    if inputFileType == 'PDB':
        inputFile = pdbFile
    elif inputFileType == 'XPLOR':
        inputFile = xplorFile


    credentials = [(FORM_USER_ID, user_id), (FORM_ACCESS_KEY, access_key)]

##############################################################################################################

    if doSave:
        data = credentials + [(FORM_ACTION,FORM_ACTION_SAVE),]
        fileObj = open(inputFile, 'rb')
        files = [( FORM_UPLOAD_FILE_BASE, inputFile, fileObj.read() ),]

        result = sendRequest(rpcUrl, data, files)
        if not result:
            nTerror("Failed to save file to server")
        else:
            print "result of save request: %s" % result
            urls = getResultUrls(credentials, entryId, DEFAULT_URL)
            print "Base URL", urls[0]
            print "Results URL:", urls[1]
            print "Log URL:", urls[2]
            print "Zip URL:", urls[3]


##############################################################################################################
    files = None

    if doRun:
        data = credentials + [(FORM_ACTION,FORM_ACTION_RUN),]
        print  sendRequest(rpcUrl, data, files)

    if doStatus:
        data = credentials + [(FORM_ACTION,FORM_ACTION_STATUS),]
        print  sendRequest(rpcUrl, data, files)

    if doLog:
        data = credentials + [(FORM_ACTION,FORM_ACTION_LOG),]
        print  sendRequest(rpcUrl, data, files)

    if doPname:
        data = credentials + [(FORM_ACTION,FORM_ACTION_PROJECT_NAME),]
        print  sendRequest(rpcUrl, data, files)

    if doPurge:
        data = credentials + [(FORM_ACTION,FORM_ACTION_PURGE),]
        print  sendRequest(rpcUrl, data, files)

if __name__ == "__main__":
    cing.verbosity = verbosityDebug
    iCingRobot()

File: trunk/cing/src/cing/Scripts/iCingRobot.java
package cing.Scripts;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Random;

import org.apache.commons.httpclient.HttpClient;
import org.apache.commons.httpclient.HttpStatus;
import org.apache.commons.httpclient.methods.PostMethod;
import org.apache.commons.httpclient.methods.multipart.FilePart;
import org.apache.commons.httpclient.methods.multipart.MultipartRequestEntity;
import org.apache.commons.httpclient.methods.multipart.Part;
import org.apache.commons.httpclient.methods.multipart.StringPart;
import org.json.JSONException;
import org.json.JSONObject;

import cing.Constants;
import cing.client.Settings;

public class iCingRobot {
//	public static String default_url = Settings.DEFAULT_URL;
	// currently JFD doesn't know how to get Java working with the https secure protocols.
	public static String default_url = "http://nmr.cmbi.ru.nl"; // disable when debugged.

	public static String rpcUrl = default_url + "/" + Settings.DEFAULT_URL_PATH + "/" + Settings.SERVLET_URL;
	public static final String SEP_DIR = "/";
	private static final int IDX_ACTION = 0;
	private static final int IDX_USER_ID = 1;
	private static final int IDX_ACCESS_KEY = 2;

	static {
		System.out.println("rpcUrl: " + rpcUrl);
	}

	public static String getNewAccessKey() {
		String allowedCharacters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
		String result = "";
		Random random = new Random();
		for (int i = 1; i <= Settings.accessKeyLength; i++) {
			int idxChar = random.nextInt(allowedCharacters.length());
			result += allowedCharacters.charAt(idxChar);
		}
		return result;
	}

	public static String join(String sep, String[] sal) {
		StringBuffer sb = new StringBuffer();
		for (int i = 0; i < sal.length; i++) {
			sb.append(sep);
			sb.append(sal[i]);
		}
		String result = sb.toString();
		if (sal.length == 0) {
			result = result.substring(1);
		}
		return result;
	}

	public static String[] getResultUrls(String[][] fields, String entryId, String url) {
		String userId = fields[IDX_USER_ID][1];
		String accessKey = fields[IDX_ACCESS_KEY][1];
		String resultBaseUrl = join(SEP_DIR, new String[] { url, "tmp/cing", userId, accessKey }).substring(1);
		String resultHtmlUrl = join(SEP_DIR, new String[] { resultBaseUrl, entryId + ".cing", "index.html" })
				.substring(1);
		String resultLogUrl = join(SEP_DIR, new String[] { resultBaseUrl, "cingRun.log" }).substring(1);
		String resultZipUrl = join(SEP_DIR, new String[] { resultBaseUrl, entryId + "_CING_report.zip" }).substring(1);

		return new String[] { resultBaseUrl, resultHtmlUrl, resultLogUrl, resultZipUrl };
	}

	/**
	 * Function to send form fields and files to a given URL. Returns null on
	 * error.
	 * */
	public static HashMap sendRequest(String url, String[][] fields, File[] files) {
		String jsonTxt = null;
		try {
			PostMethod post = new PostMethod(url);
			post.setRequestHeader("User-Agent", "anonymous");
			ArrayList al = new ArrayList();
			for (int k = 0; k < fields.length; k++) {
				al.add(new StringPart(fields[k][0], fields[k][1]));
			}
			if (files != null) {
				for (int i = 0; i < files.length; i++) {
					File f = files[i];
					al.add(new FilePart(f.getName(), f));
				}
			}
			Part[] parts = new Part[al.size()];
			for (int j = 0; j < parts.length; j++) {
				parts[j] = (Part) al.get(j);
				// System.out.println("DEBUG: Part is " + parts[j].toString());
			}
			System.out.println("\n\nDoing action:   [" + fields[IDX_ACTION][1] + "]");
			post.setRequestEntity(new MultipartRequestEntity(parts, post.getParams()));
			HttpClient client = new HttpClient();
			int status = client.executeMethod(post);
			jsonTxt = post.getResponseBodyAsString();
			// System.out.println("DEBUG: status:   [" + status + "]");
			if (status != HttpStatus.SC_OK) {
				System.err.println("Upload failed: " + post.getStatusLine());
				return null;
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
		HashMap result = processResponse(jsonTxt);

		String msgFailed = null;
		if (!result.containsKey(Settings.RESPONSE_EXIT_CODE)) {
			msgFailed = "Request not successful. Exit code was even not retrieved";
		}
		String responseExitcode = (String) result.get(Settings.RESPONSE_EXIT_CODE);
		if (!responseExitcode.equals(Settings.RESPONSE_EXIT_CODE_SUCCESS)) {
			msgFailed = "Request not successful. Exit code was: " + responseExitcode;
		}
		if (msgFailed != null) {
			msgFailed += " Response was: " + jsonTxt;
			msgFailed += "\nResult was:\n" + result.toString();
			System.err.println("ERROR: " + msgFailed);
			return null;
		}
		return result;
	}

	/**
	 * Convert the strings the iCingServer sends back into more general Java data structure.
	 */
	private static HashMap processResponse(String text) {
		HashMap dataDict = new HashMap();
		try {
			JSONObject jsonObject = new JSONObject(text);
	        for (Iterator i = jsonObject.keys(); i.hasNext();) {
	            String key = (String) i.next();
	            String value = jsonObject.getString(key);
	            dataDict.put(key, value);
	        }
		} catch (JSONException e) {
			e.printStackTrace();
		}
		return dataDict;
	}

	/**
	 * Return true on success.
	 *
	 * Expect errors without a server up and running.
	 */
	public boolean run() {
		System.out.println("Firing up the Java iCing robot; aka example interface to CING for workflows");

		// Queries possible; do one at a time going down the list.
		// After the run is started the status will let you know if the run is
		// finished
		// The log will show what the server is doing at any one time.
		boolean doSave = true; // Upload to iCing and show derived urls
		boolean doRun = false; // Start the run in Nijmegen. Use the other items first please.
		boolean doStatus = true; // Find out if the run finished
		boolean doLog = true; // Get the next piece of log file (may be empty)
		boolean doPname = true; // Get the project name back. This is the entryId below.
		boolean doPurge = true; // Remove data from server again.

		// User id should be a short id (<without any special chars.)
		// user_id = os.getenv("USER", "UnknownUser")
		String user_id = "iCingRobotJ";
//		 String access_key = getNewAccessKey(); // Use a different one in a production environment.
		String access_key = "123456";

		String entryId = "1brv"; // 68K, smallest for quick testing.
		String inputFile = join(SEP_DIR, new String[] { Constants.cingDirTestsData, "ccpn", entryId + ".tgz" })
				.substring(1);
		// System.out.println("DEBUG: inputFile: " + inputFile);
		String[][] fields = new String[][] { { Settings.FORM_PARM_ACTION, null },
				{ Settings.FORM_PARM_USER_ID, user_id }, { Settings.FORM_PARM_ACCESS_KEY, access_key }, };
		if (doSave) {
			fields[IDX_ACTION][1] = Settings.FORM_ACTION_SAVE;
			File file = new File(inputFile);
			File[] files = new File[] { file };
			// System.out.println("DEBUG: fields: " + Strings.toString(fields));
			HashMap result = sendRequest(rpcUrl, fields, files);
			if (result == null) {
				System.err.println("Failed to save file to server");
				return false;
			} else {
				System.out.println("result of save request: " + result.toString());
				String[] urls = getResultUrls(fields, entryId, default_url);
				System.out.println("Base URL	: " + urls[0]);
				System.out.println("Results URL	: " + urls[1]);
				System.out.println("Log URL		: " + urls[2]);
				System.out.println("Zip URL		: " + urls[3]);
			}
		}
		if (doRun) {
			fields[0][1] = Settings.FORM_ACTION_RUN;
			System.out.println(sendRequest(rpcUrl, fields, null));
		}
		if (doStatus) {
			fields[0][1] = Settings.FORM_ACTION_STATUS;
			System.out.println(sendRequest(rpcUrl, fields, null));
		}
		if (doLog) {
			fields[0][1] = Settings.FORM_ACTION_LOG;
			System.out.println(sendRequest(rpcUrl, fields, null));
		}
		if (doPname) {
			fields[0][1] = Settings.FORM_ACTION_PROJECT_NAME;
			System.out.println(sendRequest(rpcUrl, fields, null));
		}
		if (doPurge) {
			fields[0][1] = Settings.FORM_ACTION_PURGE; // First pair!
			System.out.println(sendRequest(rpcUrl, fields, null));
		}
		return true;
	}

	/**
	 * Expect the result: (doing just a save) rpcUrl:
	 * https://nmr.cmbi.ru.nl/icing/serv/iCingServlet Firing up the iCing robot;
	 * aka example interface to CING result of save request: {'Action': 'Save',
	 * 'Result': '67.44 kb', 'ExitCode': 'Success'} Base URL
	 * https://nmr.cmbi.ru.nl/tmp/cing/iCingRobot/sV3nfD Results URL:
	 * https://nmr.cmbi.ru.nl/tmp/cing/iCingRobot/sV3nfD/1brv.cing/index.html
	 * Log URL: https://nmr.cmbi.ru.nl/tmp/cing/iCingRobot/sV3nfD/cingRun.log
	 * Zip URL:
	 * https://nmr.cmbi.ru.nl/tmp/cing/iCingRobot/sV3nfD/1brv_CING_report.zip
	 *
	 * @param args
	 */
	public static void main(String[] args) {
		iCingRobot iRobot = new iCingRobot();
		iRobot.run();
	}
}